LNK-3737: Cleanup Docker Compose Files and Enhance Backend E2E Testing Framework

seanmcilvenna opened the pull request:
### 🛠️ Description of Changes

* Removing extra docker-compose files
* Renaming `E2ETests` to `BackendE2ETests`.
* Renaming the current `SmokeTest` to `AdhocReportingSmokeTest`, assigning it a category of `SmokeTest`.
* Updating GitHub smoke test action to only run `SmokeTest` categorized tests from the backend tests project.
* Adding authentication configuration and helper to acquire token.
* Using authentication (if configured) in smoke test execution
* Using authentication (if configured) in loading the FHIR server with data.
* Renamed `MeasureBundleLocation` to `AdhocReportingSmokeTestMeasureBundleLocation`

### 🧪 Testing Performed

* Performed E2E smoke test.
* Stepped through authentication logic during development to ensure that a token could be acquired from auth server to pass along to AdminBFF and/or FHIR Server.

### 🧑‍🔬 Unit Testing

I didn't write explicit unit tests because this is all for the purposes of "testing".

### 📓 Documentation Updated

* Describes test configuration


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced backend end-to-end tests with improved authentication support and flexible configuration via environment variables.
  - Added comprehensive documentation for backend E2E tests setup and execution.

- **Refactor**
  - Renamed the E2E test project and main test class for clarity.
  - Centralized configuration and authentication logic for backend E2E tests.
  - Improved code structure for test data loading and authentication handling.

- **Chores**
  - Updated workflow to run only backend smoke tests and adjusted test artifact paths.
  - Updated .gitignore to exclude `.runsettings` files.
  - Cleaned up solution and project files to remove unused or redundant Docker and test configurations.
  - Adjusted Docker Compose volume mount paths for telemetry services.
  - Removed Docker Compose configurations for Kafka UI, REST proxy, observability stack, and Redis cache.
  - Removed Azure Pipelines configurations.

- **Documentation**
  - Added new README for backend E2E tests.
  - Removed outdated E2E test documentation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

seanmcilvenna committed:
Removing extra docker-compose files

seanmcilvenna committed:
Renaming `E2ETests` to `BackendE2ETests`.
Renaming the current `SmokeTest` to `AdhocReportingSmokeTest`, assigning it a category of `SmokeTest`.
Updating GitHub smoke test action to only run `SmokeTest` categorized tests from the backend tests project.

seanmcilvenna committed:
* Adding authentication configuration and helper to acquire token.
* Using authentication (if configured) in smoke test execution
* Using authentication (if configured) in loading the FHIR server with data.
* Updating documentation to describe test configuration
* Renamed `MeasureBundleLocation` to `AdhocReportingSmokeTestMeasureBundleLocation`

smailliwcs reviewed (changes requested):
The "Compose" files in the project root are actually config files for OpenTelemetry, etc.  For example:

https://github.com/lantanagroup/link-cloud/blob/f7c4f9b65fef7da60eafbeeaaade5b67fc0b3ee8/docker-compose.yml#L163

So can't get rid of them, but probably appropriate to rename them ... maybe moving them into a subdirectory with all the other supporting files that `compose up` requires:

  - create-dbs.sql
  - docker-compose.{service}.yml (renamed to just {service}.yml?)
  - healthcheck-1.0.0.jar
  - kafka-client.properties
  - topics.txt

seanmcilvenna committed:
Fixing broken build by restoring the docker compose config files and moving them to a sub-directory.

seanmcilvenna committed:
Moving two more docker-specific config files to the `.docker` folder

seanmcilvenna commented:
@smailliwcs  Yup. Good point. I've moved those files to a `.docker/` sub-directory and updated the `docker-compose.yml` to reference them in their new location.

seanmcilvenna commented on Tests/BackendE2ETests/TestConfig.cs:
```diff
@@ -0,0 +1,45 @@
+        public string? Password { get; } = Environment.GetEnvironmentVariable($"{prefix}_BASICAUTH_PASSWORD");
+    }
+    
+    public class OAuthConfig(string prefix)
+    {
+        public bool ShouldAuthenticate { get; } = bool.Parse(Environment.GetEnvironmentVariable($"{prefix}_OAUTH_SHOULD_AUTHENTICATE") ?? "false");
+        public string? TokenEndpoint { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_TOKEN_ENDPOINT");
+        public string? ClientId { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_CLIENT_ID");
+        public string? ClientSecret { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_CLIENT_SECRET");
+        public string Scope { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_SCOPE") ?? "openid profile email";
+        public string? Username { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_USERNAME");
+        public string? Password { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_PASSWORD");
+    }
```
This validation already takes place in the AuthHelper class when a token is being requested.

seanmcilvenna commented on Tests/BackendE2ETests/TestConfig.cs:
```diff
@@ -0,0 +1,45 @@
+        return reader.ReadToEnd();
+    }
+
+    public class BasicAuthConfig(string prefix)
+    {
+        public bool ShouldAuthenticate { get; } = bool.Parse(Environment.GetEnvironmentVariable($"{prefix}_BASICAUTH_SHOULD_AUTHENTICATE") ?? "false");
+        public string? Username { get; } = Environment.GetEnvironmentVariable($"{prefix}_BASICAUTH_USERNAME");
+        public string? Password { get; } = Environment.GetEnvironmentVariable($"{prefix}_BASICAUTH_PASSWORD");
+    }
```
Validation of the config already takes place in the AuthHelper class.

seanmcilvenna commented on Tests/BackendE2ETests/TestConfig.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    public static string ExternalFhirServerBase => Environment.GetEnvironmentVariable("EXTERNAL_FHIR_SERVER_BASE_URL") ?? "http://localhost:6157/fhir";
+    public static string InternalFhirServerBase => Environment.GetEnvironmentVariable("INTERNAL_FHIR_SERVER_BASE_URL") ?? "http://fhir-server:8080/fhir";
+    public static string AdminBffBase => Environment.GetEnvironmentVariable("ADMIN_BFF_BASE_URL") ?? "http://localhost:8063/api";
+    public static string AdhocReportingSmokeTestMeasureBundleLocation => Environment.GetEnvironmentVariable("ADHOC_REPORTING_SMOKE_TEST_MEASURE_BUNDLE_PATH") ?? "resource://LantanaGroup.Link.Tests.E2ETests.measures.NHSNdQMAcuteCareHospitalInitialPopulation.json";
+    public static OAuthConfig AdminBffOAuth => new("ADMINBFF");
+    public static OAuthConfig FhirServerOAuth => new("FHIRSERVER");
+    public static BasicAuthConfig FhirServerBasicAuth => new("FHIRSERVER");
+
+    public static string GetEmbeddedResourceContent(string resourceName)
+    {
+        var assembly = Assembly.GetExecutingAssembly();
+        using var stream = assembly.GetManifestResourceStream(resourceName);
+        if (stream == null)
+        {
+            throw new FileNotFoundException($"Embedded resource '{resourceName}' not found.");
+        }
+
+        using var reader = new StreamReader(stream);
+        return reader.ReadToEnd();
+    }
+
+    public class BasicAuthConfig(string prefix)
+    {
+        public bool ShouldAuthenticate { get; } = bool.Parse(Environment.GetEnvironmentVariable($"{prefix}_BASICAUTH_SHOULD_AUTHENTICATE") ?? "false");
+        public string? Username { get; } = Environment.GetEnvironmentVariable($"{prefix}_BASICAUTH_USERNAME");
+        public string? Password { get; } = Environment.GetEnvironmentVariable($"{prefix}_BASICAUTH_PASSWORD");
+    }
+    
+    public class OAuthConfig(string prefix)
+    {
+        public bool ShouldAuthenticate { get; } = bool.Parse(Environment.GetEnvironmentVariable($"{prefix}_OAUTH_SHOULD_AUTHENTICATE") ?? "false");
+        public string? TokenEndpoint { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_TOKEN_ENDPOINT");
+        public string? ClientId { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_CLIENT_ID");
+        public string? ClientSecret { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_CLIENT_SECRET");
+        public string Scope { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_SCOPE") ?? "openid profile email";
+        public string? Username { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_USERNAME");
+        public string? Password { get; } = Environment.GetEnvironmentVariable($"{prefix}_OAUTH_PASSWORD");
+    }
+}
```
This entire project is for testing purposes, and creating a unit test for this would imply having a `BackendE2ETestsTests` project. Since this project as a whole is focused on testing, I don't want to test the tests, at this time.

seanmcilvenna committed:
Merge branch 'dev' into LNK-3737-enhance-e2e-tests

seanmcilvenna committed:
Removing incorrectly placed azure-pipeline files from the root folder. The `account` pipeline already exists in `Azure_Pipelines/` folder, as well.

seanmcilvenna committed:
Merge remote-tracking branch 'origin/LNK-3737-enhance-e2e-tests' into LNK-3737-enhance-e2e-tests

edward-miller-lcg committed:
Merge branch 'dev' into LNK-3737-enhance-e2e-tests
